src/pexprs-outputRecipe.js

```diff
@@ -36,91 +28,95 @@ function getIntervalInfo(expr, grammarInterval) {
+pexprs.Extend.prototype.outputRecipe = function(formals, grammarInterval) {
   var extension = this.terms[0]; // [extension, orginal]
-  extension.outputRecipe(sb, formals, grammarInterval);
+  return extension.outputRecipe(formals, grammarInterval);
 };
 
-pexprs.Seq.prototype.outputRecipe = function(sb, formals, grammarInterval) {
-  sb.append('this.seq(');
-  for (var idx = 0; idx < this.factors.length; idx++) {
-    if (idx > 0) {
-      sb.append(', ');
-    }
-    this.factors[idx].outputRecipe(sb, formals, grammarInterval);
-  }
-  sb.append(')' + getIntervalInfo(this, grammarInterval));
+pexprs.Seq.prototype.outputRecipe = function(formals, grammarInterval) {
+  return [
+    'seq',
+    getMetaInfo(this, grammarInterval)
+  ].concat(this.factors.map(function(factor) {
+    return factor.outputRecipe(formals, grammarInterval);
+  }));
 };
 
 pexprs.Star.prototype.outputRecipe =
 pexprs.Plus.prototype.outputRecipe =
 pexprs.Opt.prototype.outputRecipe =
 pexprs.Not.prototype.outputRecipe =
-pexprs.Lex.prototype.outputRecipe = function(sb, formals, grammarInterval) {
-  sb.append('this.' + this.constructor.name.toLowerCase() + '(');
-  this.expr.outputRecipe(sb, formals, grammarInterval);
-  sb.append(')' + getIntervalInfo(this, grammarInterval));
+pexprs.Lex.prototype.outputRecipe = function(formals, grammarInterval) {
+  return [
+    this.constructor.name.toLowerCase(),
+    getMetaInfo(this, grammarInterval),
+    this.expr.outputRecipe(formals, grammarInterval)
+  ];
 };
 
-pexprs.Lookahead.prototype.outputRecipe = function(sb, formals, grammarInterval) {
-  sb.append('this.la(');
-  this.expr.outputRecipe(sb, formals, grammarInterval);
-  sb.append(')' + getIntervalInfo(this, grammarInterval));
+pexprs.Lookahead.prototype.outputRecipe = function(formals, grammarInterval) {
+  return [
+    'la',
+    getMetaInfo(this, grammarInterval),
+    this.expr.outputRecipe(formals, grammarInterval)
+  ];
 };
 
-pexprs.Apply.prototype.outputRecipe = function(sb, formals, grammarInterval) {
-  sb.append('this.app(');
-  sb.append(JSON.stringify(this.ruleName));
-  if (this.ruleName.indexOf('_') >= 0 && formals.length > 0) {
-    var apps = formals.
-        map(function(_, idx) { return 'this.param(' + idx + ')'; });
-    sb.append(', [' + apps.join(', ') + ']');
-  } else if (this.args.length > 0) {
-    sb.append(', [');
-    this.args.forEach(function(arg, idx) {
-      if (idx > 0) {
-        sb.append(', ');
-      }
-      arg.outputRecipe(sb, formals, grammarInterval);
-    });
-    sb.append(']');
-  }
-  sb.append(')' + getIntervalInfo(this, grammarInterval));
+pexprs.Apply.prototype.outputRecipe = function(formals, grammarInterval) {
+  return [
+    'app',
+    getMetaInfo(this, grammarInterval),
+    this.ruleName,
+    this.args.map(function(arg) {
+      return arg.outputRecipe(formals, grammarInterval);
+    })
+  ];
+  // TODO: fold this into the main recipe creation (since _ indicates case)
```

Comment:
What's currently still missing, is that the rule-level alteration,
```
rule = sub1 -- one
     | sub2 -- two
```
gets folded into the rule definition. Just as with the old recipe format, those inline rules get serialized separately:
```json
{
  "rule_one": ["define",{},null,[],["app",{},"sub1",[]]],
  "rule_two": ["define",{},null,[],["app",{},"sub2",[]]],
  "rule": ["define",{},null,[],
    ["alt",{},
      ["app",{},"rule_one",[]],
      ["app",{},"rule_two",[]]
    ]
  ]
}
```
It would be possible to fold that to something like:
```json
{
  "rule": ["define",{},null,[],
    ["app",{"case":"one"},"sub1",[]],
    ["app",{"case":"two"},"sub2",[]]
  ]
}
```
Preferences? Opinions? @alexwarth @pdubroy